### DISCUSSION TITLE
Undo file added to git-lfs

### DISCUSSION THREAD

**Discussion by siddhsql** [git-lfs/git-lfs#5795] [General]
In course of using git-lfs I ended up adding a file to LFS that shouldn't have been and it got pushed to github as well. How can I undo this now? i.e., remove the file from LFS and add it to normal git. I stumbled upon [this](https://docs.github.com/en/repositories/working-with-files/managing-large-files/removing-files-from-git-large-file-storage) but the first step here seems complicated and dangerous to me:

>Remove the file from the repository's Git history using either the filter-repo command or BFG Repo-Cleaner. For detailed information on using these, see "[Removing sensitive data from a repository](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/removing-sensitive-data-from-a-repository)."

>When altering your repository's history using tools like git filter-repo or the BFG Repo-Cleaner, it's crucial to understand the implications, especially regarding open pull requests and sensitive data.

I don't understand it very well. Can someone help me with this? I was hoping for a series of commands that modify my local repo first and then I push those changes to remote. Whereas the steps in those links seem to modify remote directly bypassing the local which is something I have never done in my course of using git. With git, I always make changes to my local first, then push those to remote.

**Comment by siddhsql** [comment]
I will answer the question myself (also to make a note for self). The procedure that worked was to:

```
git rm the file
git commit
git push
```

this removes the file from LFS

then edit `.gitattributes` so that the file will not be tracked by git lfs

and then add the file back again:

```
git add the file
git commit
git push
```

this time the file will be added to normal git provided you edited .gitattributes appropriately

**Comment by chrisd8088** [comment]
I'm glad you found something which works for you!

One thing I might add is that the approach you've taken likely leaves the `.gitattributes` file with the incorrect pattern, and the Git LFS file(s) you didn't want, in your Git commit history, just not at the tip of your current branch.

If you want to rewrite that branch to get rid of those older commits, regular Git commands like `git rebase -i` and `git revert` can help you remove or edit previous commits.  You might then need to "force push" your branch (`git push -f`) to your remote so it has the rewritten history too; note that this is a "force" because it overwrites the previous branch history on the remotes.  There's lots of documentation about these regular Git commands if you're interested.

There's also the `git lfs migrate export` command, which can automatically rewrite a branch or branches for you, removing files from Git LFS and adjusting `.gitattributes` appropriately.  So if you accidentally added a rule for `*.xyz` files and want to undo that and all the matching files which were converted into Git LFS objects, `git lfs migrate export --include="*.xyz"` should do that for you (on your current branch; you can also specify multiple branches, or `--everything` for all branches).  Again, this will change your local Git history, so you would then want to force-push your branch to your remote, assuming things look they way you want them.

---

For reference, another detail to consider is that, once you've push a Git LFS object to a remote, the service may not remove it even if you force-push your branches and remove it from your Git history entirely.  That's up to the individual service's implementation.  GitHub, for instance, will only remove your Git LFS objects once you remove the repository (and any forks) entirely, so if you're concerned about your consumption of Git LFS storage, you should check your Git hosting service's support documentation.

In this case, though, it sounds like you only had a few accidentally converted Git LFS objects, so there's probably nothing to worry about in terms of overall storage consumption.  But, if you had many thousands of large files and wanted to remove them from Git LFS so you weren't charged for their storage, this is could be a concern.
